File: 121 BestTimeBuySellStock/dp.py
from typing import List


class Solution:
    def maxProfit(self, prices:List[int]) -> int:
        
        min_price, max_profit = float('inf'), 0

        for price in prices:
            
            if min_price > price:
                min_price = price
            profit = price - min_price

            if profit > max_profit:
                max_profit = profit

        return max_profit

print(Solution().maxProfit([7,1,5,3,6,4])) # solution: 5
print(Solution().maxProfit([7,6,4,3,1])) # solution: 0

File: DSA/1 arrays/buySellStock/dp.cpp
using namespace std;
#include <vector>
#include <algorithm>
#include <iostream>

class Solution {
    public: 
    int maxProfit(vector<int> & prices) {

        int minPrice = prices[0];
        int maxProfit = 0;
        int n = prices.size();

        for(int i = 1; i < n; i++){

            if (prices[i] < minPrice) 
                minPrice = prices[i];
            
            maxProfit = std::max(maxProfit, prices[i] - minPrice);
        }
        return maxProfit;

    }
};

int main() {
    Solution sol;
    vector<int> prices = {7, 1, 5, 3, 6, 4};
    cout << "Max Profit : " << sol.maxProfit(prices) << endl;
    return 0;

}